src/test/scala/vsys/blockchain/state/contract/channel/PaymentChannelContractDiffTest.scala

```diff
@@ -72,35 +84,357 @@ class PaymentChannelContractDiffTest extends PropSpec
+    updateTimeType = Seq(DataType.ShortBytes, DataType.Timestamp)
+    updateTime <- updateExpiredTimeChannelGen(master, contractId, updateTimeData, updateTimeType, attach, fee, ts)
+    executeWithdraw <- executeWithdrawChannelGen(master, contractId, create.id.arr, attach, fee, ts)
+    terminateChannel <- terminateChannelGen(master, contractId, create.id.arr, attach, fee, ts)
+  } yield (genesis, regContract, depositVSYS, create, executeWithdraw, updateTime, terminateChannel, fee, ts)
+
+  property("execute update expired time successfully") {
+    forAll(preconditionsAndPaymentChannelFunctionTest) { case (genesis: GenesisTransaction, reg: RegisterContractTransaction,
+      deposit: ExecuteContractFunctionTransaction, create: ExecuteContractFunctionTransaction, _,
+      updateTime: ExecuteContractFunctionTransaction, _, _, _) =>
+      assertDiffEi(Seq(TestBlock.create(Seq(genesis)), TestBlock.create(Seq(reg, deposit, create))),
+        TestBlock.create(Seq(updateTime))) { blockDiffEi =>
+        blockDiffEi shouldBe an[Right[_, _]]
+        blockDiffEi.explicitGet().txsDiff.contractDB.isEmpty shouldBe false
+        blockDiffEi.explicitGet().txsDiff.contractNumDB.isEmpty shouldBe true
+        blockDiffEi.explicitGet().txsDiff.portfolios.isEmpty shouldBe false
+        blockDiffEi.explicitGet().txsDiff.txStatus shouldBe TransactionStatus.Success
+      }
+    }
+  }
+
+  property("terminate channel successfully") {
+    forAll(preconditionsAndPaymentChannelFunctionTest) { case (genesis: GenesisTransaction, reg: RegisterContractTransaction,
+      deposit: ExecuteContractFunctionTransaction, create: ExecuteContractFunctionTransaction,
+      _, _, terminate: ExecuteContractFunctionTransaction, _, _) =>
+        assertDiffEi(Seq(TestBlock.create(Seq(genesis)), TestBlock.create(Seq(reg, deposit, create))),
+          TestBlock.create(Seq(terminate))) { blockDiffEi =>
+          blockDiffEi shouldBe an[Right[_, _]]
+          blockDiffEi.explicitGet().txsDiff.contractDB.isEmpty shouldBe false
+          blockDiffEi.explicitGet().txsDiff.contractNumDB.isEmpty shouldBe true
+          blockDiffEi.explicitGet().txsDiff.portfolios.isEmpty shouldBe false
+          blockDiffEi.explicitGet().txsDiff.txStatus shouldBe TransactionStatus.Success
+        }
+    }
+  }
+
+  property("execute withdraw able to withdraw funds from channel after channel expires") {
+    forAll(preconditionsAndPaymentChannelFunctionTest) { case (genesis: GenesisTransaction, reg: RegisterContractTransaction,
+      deposit: ExecuteContractFunctionTransaction, create: ExecuteContractFunctionTransaction,
+      executeWithdraw: ExecuteContractFunctionTransaction, _, _, fee: Long, ts: Long) =>
+        assertDiffAndStateCorrectBlockTime(Seq(TestBlock.create(genesis.timestamp, Seq(genesis)), TestBlock.create(create.timestamp, Seq(reg, deposit, create))),
+          TestBlock.create(create.timestamp + 1000000000001L, Seq(executeWithdraw))) { (blockDiff, newState) =>
```

Comment:
I am not sure this is a valid case. since load timestamp opc will return the last block's timestamp. which means  change the timestamp in the current block is useless